app/src/main/proto/vss.proto

```diff
@@ -0,0 +1,169 @@
+syntax = "proto3";
+option java_multiple_files = true;
+package org.vss;
+
+/*
+  StoreId: All APIs operate within a single storeId.
+  StoreId is keyspace identifier.
+  Ref: https://en.wikipedia.org/wiki/Keyspace_(distributed_data_store)
+  It is up to clients to use single or multiple store's for their usage.
+  This can be used for client-isolation/ rate-limiting / throttling on server-side.
+  Authorization and billing can also be performed at storeId level.
+ */
+message GetObjectRequest {
+
+  string storeId = 1;
+
+  /*
+    Key for which the value is to be fetched.
+
+    Consistency Guarantee:
+    Get/Read operations against a key are consistent reads and will reflect all previous writes,
+    since Put/Write provides read-after-write and read-after-update consistency guarantees.
+
+    Read Isolation:
+    Get/Read operations against a key are ensured to have read-committed isolation.
+    Read-committed is an isolation level that guarantees that any data read was committed at the
+    moment it is read. It will not return with any intermediate, uncommitted or 'dirty' read while
+    another PutObjectRequest was concurrently updating the same keys.
+    Read-committed isolation does not prevent modifications of the key-value immediately after the
+    read.
+   */
+  string key = 2;
+}
+
+
+message GetObjectResponse {
+  /*
+    Fetched value and version along with corresponding key in request.
+   */
+  KeyValue value = 2;
+}
+
+
+message PutObjectRequest {
+
+  string storeId = 1;
+  /*
+    If present, write will only succeed if current DB globalVersion against a storeId is
+    same as in request.
+    Client is expected to store(client-side) global version against storeId.
+    And PutObjectRequest should contain their client-side value of globalVersion.
+
+    For first write of store, global version should be '0'. If write succeeds, clients must increment
+    their global version(client-side) by 1. This updated global version should be used in subsequent
+    PutObjectRequests for the store.
```

Comment:
Server does increment the value(but only server-side) on every put, hence we need to increment value client-side for subsequent writes. 
Increment is required to ensure that value is same across client and server. 
I didn't understand how this conflicts with first paragraph.

Reply:
> I didn't understand how this conflicts with first paragraph.

The first paragraph states the write will only succeed if the versions are the same. But if the client increments the version, then the next write would fail with a naive reading of the first paragraph. But it actually won't, as you pointed out, since the server increments the version which wasn't clear.

It seems more intuitive if the client incremented prior to writing rather than after writing. Otherwise, the client's idea of the global version would be one ahead of the actual one. I guess it depends on how you look at it, especially given the empty case when 0 should be used.

Or better yet, I suppose we could always have the server return the global version in `PutObjectResponse` on success and dispense entirely with the complexity around having the client increment. Likewise when fetching all data. Thoughts? Could even be an opaque type, if desirable.

Reply:
Added more clarification in most recent commit for the same.

The reason client-side increment is after write is because write can fail. 
In increment-before-write approach, if write fails we will have to rollback local/secondary version numbers which could get messy to keep track of.
We don't want to be in situation where local/secondary store is ahead of primary storage. primary ahead of secondary is acceptable scenario. 

Client-side increment is common pattern used for optimistic locking. Regarding global version in response, there are key-versions as well that we need to increment and keep track of. 
One concern with client using version numbers directly from server response, is that it introduces additional trust requirement on server. 
If client blindly applies version numbers from response, it could introduce additional risk/attack-vector. (note that this risk is still present but limited to when we are syncing whole data. And we can add additional verifications there.)
That's why i thought it would be better to let client apply version increments independently. (whether it is used along with single or multiple servers)

Reply:
> If client blindly applies version numbers from response, it could introduce additional risk/attack-vector. (note that this risk is still present but limited to when we are syncing whole data. And we can add additional verifications there.)

What sort of verification? Checking that the version numbers returned are greater than or equal to the local versions? Or something more?

Reply:
In PutResponse, I mean to check correctness one way is for client to run the exact same logic of increment. Which renders the server-responded versions pointless. 

And by returning versions in response, we don't want to encourage blindly applying those version-numbers on client-side as a practice.

In case of applying versions while syncing full data, one possibility is to verify against local version(>=) and also with version numbers from another server(==).